neps/nep-0605-sharded-contracts.md

```diff
@@ -0,0 +1,477 @@
+
+### Requirements
+
+We can now explain what the high level requirements are that the above proposal is trying to meet.  This provides the sufficient context to understand why the proposal is designed the way it is.
+
+#### Multiple contract codes
+
+It should be possible to use multiple sharded contract codes on a single account.  This is important because otherwise users will have to create new accounts for each type of FT they want to hold which means that the users will have to manage multiple keys, etc. degrading user experience.
+
+We considered the existing subaccount feature as well here.  That idea does not work because subaccounts only provide access control over account creation.  The keys for each subaccount still need to be managed individually.
+
+#### Access control to sharded contract functions
+
+We identified that sharded contracts may want to have 2 types of access control on functions.
+
+First are functions that can only be called by the account owner (e.g. `send_tokens()`).  This scenario is covered by the existing set of host functions.
+
+Second are functions that can only be called by another instance of the same contract.  Here the `ShardedFunctionCallAction` action provides information about the signer and the runtime can provide information about the current account.  And then as seen in `receive_tokens()`, the contract can then perform the appropriate checks.
+
+#### Storage namespaces
+
+In the pseudocode, we see the contract is storing its state locally on the accounts.  Without additional primitives, malicious users could tamper with this state.  In the FT example above, a malicious user could have initialised the FT balance on its state to a high value before deploying the sharded contract on the account which would allow it to malicious mint tokens.
+
+Further, if we have multiple contract codes running on the same account, they could accidentally overwrite or modify another contract code state.  E.g. if two different sharded FT contracts are deployed on a single account and they both want to make changes to the `balance` key.
+
+Hence, we need a way to enable storage namespaces which guarantees that whatever state a sharded contract code is storing on an account, it cannot be modified by another contract code on that account.
+
+### Detailed specification
+
+With the high level requirements and the pseudocode presented, we can discuss the specification of the new primitives been proposed.
+
+#### `DeployShardedContractAction`
+
+This action is similar to the `DeployGlobalContractAction`.  Note that an account can have only a single sharded contract code deployed on it.  Processing this action will generate a `GlobalContractDistributionReceipt`.  We propose the following modifications to this receipt.
```

Comment:
I think we should explore if we can make global vs local contract code completely orthogonal to sharded contracts. I feel it might be nicer.

We could probably get away without touching global contract distribution code. I think when a user uses `UseShardedContractAction`, all the network has to check is if there is a global contract with that contract identifier. That shouldn't require anything special on the code distribution side.

In that case, the difference between a global sharded contract and a global non-sharded contract is simply whether the user has submitted `UseShardedContractAction` or not.

Reply:
I agree, if we could simplify that, that will be ideal.  The challenge is distributing the `version`.  

The more I think about it, the more I feel like `version` is adding a lot of additional complexity.  

One way, contracts could handle `version` entirely in the application space is by adding the version to the function names e.g. `receive_tokens_v1()`.

So when a sender sends a message, they are sending it to a specific version: `receive_tokens()_vX`.  And when a receiver upgrades, it is "rejecting" messages sent by earlier versions of the sender by replacing `receive_tokens_v[X]` by `receive_tokens_v[X+1]`.  

Do you think, the above would be good enough for apps to use and then we can kill the whole `version` business?

Reply:
Nice idea. I believe such version naming convention would work, yeah. Might be more effective than having versions be part of the protocol.

The only downside I see is that we basically push the problem to developers, without much help. Solving it in the SDK in such a way might be possible but sounds quite hacky...

I'll try to think of alternative solutions before our next call.

Reply:
Also, this naming convention wouldn't protect from malicious code swap, right?
In that case, anytime a sharded contract uses a global code by account id, it has to fully trust the owner of that account, I guess.